### PR TITLE
Fix errors and warning related to derived variables

### PR DESCRIPTION
**Pull request by anagainaru**
PR to fix errors and warnings related to derived variables (except the lexer related warnings since @lizdulac is opening a PR on fixing lexer issues)

## Error

**Mismatch in reading count when reading variables that are part of a derived variable**

`Req->Count.data()` are the dimensions of the derived variable in the current request

`VarRec->DimCount` are the total dimensions of one of the variables used by the derived variable.

In all ecause Curl adds one dimension at the end).

However, I am adding now an aggregated operator that will have a lower dimension than the composing variables, so this line seg faults.

### CODE REVIEW

**Comment by anagainaru** [comment]
Also @eisenhauer I am very disturbed that we use VarRec for two reasons in the same file (`BP5Deserializer.cpp`)
```c++
Line 1574:         auto VarRec = (struct BP5VarRec *)Req->VarRec;
Line 1730:                                     BP5VarRec *VarRec = VarByName.at(varBase->m_Name);
```

Both in the same function (`BP5Deserializer::GenerateReadRequests(const bool doAllocTempBuffers, size_t *maxReadSize)`) 

It makes everything hard to read, should we rename the variable on line 1730 to `VarPrimaryRec` or some name to indicate that this is the rec of the composing variable?

**Comment by eisenhauer** [comment]
> Also @eisenhauer I am very disturbed that we use VarRec for two reasons in the same function (in `BP5Deserializer.cpp`)
> 
> ```c++
> Line 1574:         auto VarRec = (struct BP5VarRec *)Req->VarRec;
> Line 1730:                                     BP5VarRec *VarRec = VarByName.at(varBase->m_Name);
> ```

I completely agree.  Actually this is my fault for not trying this in CI with ADIOS2_USE_Derived_Variable set to on everywhere.  At least one of the compilers in our CI would have warned about the nested variable hiding the more global one.

**Comment by anagainaru** [comment]
I compiled your PR before approving but didn't see the warning either :) so both of our faults.

I'll change the name and push over this tonight. I'll wait until Norbert's PR gets merged so I can rebase and not have all these failures.

**Comment by anagainaru** [comment]
This PR can be reviewed

File: source/adios2/toolkit/derived/Function.cpp
Comment (eisenhauer) [inline source/adios2/toolkit/derived/Function.cpp:115]: Not that this is wrong, but curious about why the change from PRIMITIVE_STDTYPE to PRIMITIVE_ATTRIBUTE_STDTYPE.   This just drops the complex types, from being function candidates.  Is there some problem with allowing them?

Comment (anagainaru) [reply source/adios2/toolkit/derived/Function.cpp:115]: It is temporary until we decide how to allocate memory for the derived variables. Right now, I didn't want to deal with complex types when using malloc and memset. I added it as a separate commit so I can revert only this one

**Review by eisenhauer — APPROVED** [review]
Looks good then.

**Comment by vicentebolea** [comment]
This PR cannot be backported since it depends on another PR that breaks ABI/API comp #4228

**Comment by anagainaru** [comment]
This would mean that we will be able to read derived variables only from 2.11 on. For now we can just write them and/or store stats. I am fine with this. I will mark both this and #4228 for release 2.11

**Pull request merged**
